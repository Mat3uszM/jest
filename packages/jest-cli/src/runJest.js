/**
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @flow
 */
'use strict';

import type {Context} from 'types/Context';
import type {Test} from 'types/TestRunner';
import type {PatternInfo} from './SearchSource';
import type TestWatcher from './TestWatcher';

const fs = require('graceful-fs');

const SearchSource = require('./SearchSource');
const TestRunner = require('./TestRunner');
const TestSequencer = require('./TestSequencer');

const getTestPathPatternInfo = require('./lib/getTestPathPatternInfo');
const chalk = require('chalk');
const {Console, formatTestResults} = require('jest-util');
const getMaxWorkers = require('./lib/getMaxWorkers');
const path = require('path');
const setState = require('./lib/setState');

const getTestSummary = (argv: Object, patternInfo: PatternInfo) => {
  const testPathPattern = SearchSource.getTestPathPattern(patternInfo);
  const testInfo = patternInfo.onlyChanged
    ? chalk.dim(' related to changed files')
    : patternInfo.input !== '' ? chalk.dim(' matching ') + testPathPattern : '';

  const nameInfo = argv.testNamePattern
    ? chalk.dim(' with tests matching ') + `"${argv.testNamePattern}"`
    : '';

  return chalk.dim('Ran all test suites') +
    testInfo +
    nameInfo +
    chalk.dim('.');
};

const getTestPaths = async (context, patternInfo, argv, pipe) => {
  const source = new SearchSource(context, context.config);
  let data = await source.getTestPaths(patternInfo);
  if (!data.paths.length) {
    const localConsole = new Console(pipe, pipe);
    if (patternInfo.onlyChanged && data.noSCM) {
      if (context.config.watch) {
        // Run all the tests
        setState(argv, 'watchAll', {
          noSCM: true,
        });
        patternInfo = getTestPathPatternInfo(argv);
        data = await source.getTestPaths(patternInfo);
      } else {
        localConsole.log(
          'Jest can only find uncommitted changed files in a git or hg ' +
            'repository. If you make your project a git or hg ' +
            'repository (`git init` or `hg init`), Jest will be able ' +
            'to only run tests related to files changed since the last ' +
            'commit.',
        );
      }
    }

    localConsole.log(
      source.getNoTestsFoundMessage(patternInfo, context.config, data),
    );
  }

  return {
    data,
    patternInfo,
  };
};

const runJest = async (
  contexts: Array<Context>,
  argv: Object,
  pipe: stream$Writable | tty$WriteStream,
  testWatcher: TestWatcher,
  startRun: () => *,
  onComplete: (testResults: any) => void,
) => {
  const maxWorkers = getMaxWorkers(argv);
  const context = contexts[0];
  const testRunData = await Promise.all(
    contexts.map(async context => {
      const {data, patternInfo} = await getTestPaths(
        context,
        getTestPathPatternInfo(argv),
        argv,
        pipe,
      );

      const sequencer = new TestSequencer(context);
      const tests = sequencer.sort(data.paths);
      return {context, patternInfo, sequencer, tests};
    }),
  );

  const processResults = runResults => {
    if (context.config.testResultsProcessor) {
      /* $FlowFixMe */
      runResults = require(context.config.testResultsProcessor)(runResults);
    }
    if (argv.json) {
      if (argv.outputFile) {
        const outputFile = path.resolve(process.cwd(), argv.outputFile);

        fs.writeFileSync(
          outputFile,
          JSON.stringify(formatTestResults(runResults)),
        );
        process.stdout.write(
          `Test results written to: ` +
            `${path.relative(process.cwd(), outputFile)}\n`,
        );
      } else {
        process.stdout.write(JSON.stringify(formatTestResults(runResults)));
      }
    }
    return onComplete && onComplete(runResults);
  };

  const allTests = testRunData
    .reduce((tests, testRun) => tests.concat(testRun.tests), [])
    .sort((a: Test, b: Test) => {
      if (a.duration != null && b.duration != null) {
        return a.duration < b.duration ? 1 : -1;
      }
      return a.duration == null ? 1 : 0;
    });

  if (
    allTests.length === 1 &&
    context.config.silent !== true &&
    context.config.verbose !== false
  ) {
    context.config = Object.assign({}, context.config, {verbose: true});
  }

  if (context.config.updateSnapshot === true) {
     context.config = Object.assign({}, context.config, {
      updateSnapshot: true,
    });
  }

  const results = await new TestRunner(
    context,
    context.config,
    {
      getTestSummary: () => getTestSummary(argv, testRunData[0].patternInfo),
      maxWorkers,
    },
    startRun,
  ).runTests(allTests, testWatcher);
  testRunData.forEach(({sequencer, tests}) =>
    sequencer.cacheResults(tests, results));
  return processResults(results);
};

module.exports = runJest;
