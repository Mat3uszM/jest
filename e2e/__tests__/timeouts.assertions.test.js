/**
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * @flow
 */

'use strict';

import path from 'path';
import {cleanup, extractSummary, writeFiles} from '../Utils';
import {skipSuiteOnJasmine} from '../../scripts/ConditionalTest';

import runJest from '../runJest';

const DIR = path.resolve(__dirname, '../timeouts');

skipSuiteOnJasmine();
beforeEach(() => cleanup(DIR));
afterAll(() => cleanup(DIR));

test('exceeds the timeout, without reporting expect.assertions', () => {
  writeFiles(DIR, {
    '__tests__/a-banana.js': `
      jest.setTimeout(20);

      test('banana', () => {
        expect.assertions(2);
        return new Promise(resolve => {
          setTimeout(resolve, 100);
        });
      });
    `,
    'package.json': '{}',
  });

  const {stderr, status} = runJest(DIR, ['-w=1', '--ci=false']);
  const {rest, summary} = extractSummary(stderr);
  expect(rest).toMatch(
    /(jest\.setTimeout|jasmine\.DEFAULT_TIMEOUT_INTERVAL|Exceeded timeout)/,
  );
  expect(rest).not.toMatch(/Expected two assertions/);
  expect(summary).toMatchSnapshot();
  expect(status).toBe(1);
});

test('exceeds the timeout synchronously', () => {
  writeFiles(DIR, {
    '__tests__/a-banana.js': `
      jest.setTimeout(20);

      test('banana', () => {
        expect.assertions(2);
        const startTime = Date.now();
        while (Date.now() - startTime < 100) {
        }
      });
    `,
    'package.json': '{}',
  });

  const {stderr, status} = runJest(DIR, ['-w=1', '--ci=false']);
  const {rest, summary} = extractSummary(stderr);
  expect(rest).toMatch(
    /(jest\.setTimeout|jasmine\.DEFAULT_TIMEOUT_INTERVAL|Exceeded timeout)/,
  );
  expect(rest).not.toMatch(/Expected two assertions/);
  expect(summary).toMatchSnapshot();
  expect(status).toBe(1);
});
