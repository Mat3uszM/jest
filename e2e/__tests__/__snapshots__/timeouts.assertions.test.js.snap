// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exceeds the timeout synchronously 1`] = `
"FAIL __tests__/a-banana.js
  ✕ banana

  ● banana

    Exceeded timeout of 20ms for a test.
    Use jest.setTimeout(newTimeout) to increase the timeout value, if this is a long-running test.

      at packages/jest-circus/build/utils.js:264:15

"
`;

exports[`exceeds the timeout synchronously 2`] = `
"Test Suites: 1 failed, 1 total
Tests:       1 failed, 1 total
Snapshots:   0 total
Time:        <<REPLACED>>
Ran all test suites.
"
`;

exports[`exceeds the timeout, without reporting expect.assertions 1`] = `
"FAIL __tests__/a-banana.js
  ✕ banana

  ● banana

    thrown: \\"Exceeded timeout of 20ms for a test.
    Use jest.setTimeout(newTimeout) to increase the timeout value, if this is a long-running test.\\"

      2 |       jest.setTimeout(20);
      3 | 
    > 4 |       test('banana', () => {
        |       ^
      5 |         expect.assertions(2);
      6 |         return new Promise(resolve => {
      7 |           setTimeout(resolve, 100);

      at __tests__/a-banana.js:4:7

"
`;

exports[`exceeds the timeout, without reporting expect.assertions 2`] = `
"Test Suites: 1 failed, 1 total
Tests:       1 failed, 1 total
Snapshots:   0 total
Time:        <<REPLACED>>
Ran all test suites.
"
`;
